pandas/core/frame.py

```diff
@@ -1941,7 +1941,11 @@ def _getitem_array(self, key):
                 warnings.warn("Boolean Series key will be reindexed to match "
                               "DataFrame index.", UserWarning)
             elif len(key) != len(self.index):
-                raise ValueError('Item wrong length %d instead of %d.' %
+                if lib.infer_dtype(self._info_axis)=="boolean":
```

Comment:
I think you can move this check to ``_convert_to_indexer``, and raise if their is a problem (e.g. subsume the ``ValueError`` there). makes this code a lot simpler. right now too much special casing.